src/TMap.cpp

```diff
@@ -2649,6 +2653,550 @@ void TMap::setRoomNamesShown(bool shown)
+    mpProgressDialog->close();
+    mpProgressDialog = nullptr;
+
+    return {file.error() == QFileDevice::NoError,
+                ((file.error() == QFileDevice::NoError) ? QString() : QStringLiteral("could not export file, reason: %1").arg(file.errorString()))};
+}
+
+std::pair<bool, QString> TMap::readJsonMapFile(const QString& source)
+{
+    const QString oldDefaultAreaName{mDefaultAreaName};
+    const QString oldUnnamedName{mUnnamedAreaName};
+
+    if (mpProgressDialog) {
+        return {false, QStringLiteral("import or export already in progress")};
+    }
+
+    QFile file(source);
+    if (!file.open(QFile::ReadOnly)) {
+        qWarning().noquote().nospace() << "TMap::readJsonMapFile(...) WARNING - Could not open JSON file \"" << source << "\".";
+        return {false, QStringLiteral("could not open file \"%1\"").arg(source)};
+    }
+
+    QByteArray mapData = file.readAll();
+    file.close();
+    QJsonParseError jsonErr;
+    QJsonDocument doc(QJsonDocument::fromJson(mapData, &jsonErr));
+    if (jsonErr.error != QJsonParseError::NoError) {
+        return {false, QStringLiteral("could not parse file \"%1\", reason: \"%2\" at offset %3")
+                    .arg(source, jsonErr.errorString(), QString::number(jsonErr.offset))};
+    }
+
+    if (doc.isEmpty()) {
+        qDebug().nospace().noquote() << "TMap::readJsonMapFile(\"" << source << "\") INFO - no Json file data detected, this is not a Mudlet JSON map file.";
+        return {false, QStringLiteral("empty Json file, no map data detected")};
+    }
+
+    // Read all the base level stuff:
+    QJsonObject mapObj{doc.object()};
+    double formatVersion = 0.0f;
+    if (mapObj.contains(scFORMAT_VERSION) && mapObj[scFORMAT_VERSION].isDouble()) {
+        formatVersion = mapObj[scFORMAT_VERSION].toDouble();
+        if (qFuzzyCompare(1.0, formatVersion + 1.0) || formatVersion < 1.0000 || formatVersion > 1.0000) {
+            // We only handle 1.000f right now (0.001f was borked, 0.002f
+            // didn't include room symbol color, 0.003 is the same as 1.000
+            // but the numbered was changed for release into the wild):
+            qDebug().nospace().noquote() << "TMap::readJsonMapFile(\"" << source << "\") INFO - Version information was found: " << formatVersion << "and it is not okay.";
+            return {false, QStringLiteral("invalid version: %1 detected").arg(formatVersion, 0, 'f', 3, QLatin1Char('0'))};
+        }
+    } else {
+        qDebug().nospace().noquote() << "TMap::readJsonMapFile(\"" << source << "\") INFO - Version information was not found, this is not likely to be a Mudlet JSON map file.";
+        return {false, QStringLiteral("no version number detected")};
+    }
+
+    if (!mapObj.contains(scAREAS) || !mapObj.value(scAREAS).isArray()) {
+        return {false, QStringLiteral("no areas detected")};
+    }
+
+    mProgressDialogAreasTotal = qRound(mapObj[scAREA_COUNT].toDouble());
```

Comment:
Why all of this conversion? This is something that could use a comment

Reply:
All numbers (even integers) are `doubles` in JSON aren't and this seemed to me to be the most robust means of reading something that *could* have been edited by a user (or even a third party tool).